ddtrace/contrib/internal/asgi/middleware.py

```diff
@@ -241,9 +259,288 @@ async def __call__(self, scope, receive, send):
+                                except ValueError:
+                                    pass
+                            # set link to http handshake span
+                            send_span.set_link(
+                                trace_id=span.trace_id, span_id=span.span_id, attributes={"dd.kind": "resuming"}
+                            )
+                            send_span.set_metric("websocket.message.frames", 1)
+                            if "text" in message:
+                                send_span.set_tag_str("websocket.message.type", "text")
+                                send_span.set_metric("websocket.message.length", len(message["text"].encode("utf-8")))
+                            elif "binary" in message:
+                                send_span.set_tag_str("websocket.message.type", "binary")
+                                send_span.set_metric("websocket.message.length", len(message["bytes"]))
+
+                    elif (
+                        scope["type"] == "websocket"
+                        and self.integration_config._trace_asgi_websocket_messages
+                        and message.get("type") == "websocket.close"
+                    ):
+                        """
+                        tags:
+                        websocket.close.code
+                        websocket.close.reason
+
+                        """
+                        current_receive_span = scope.get("datadog", {}).get("current_receive_span")
+
+                        # Use receive span as parent if it exists, otherwise use main span
+                        if current_receive_span:
+                            parent_span = current_receive_span
+                        else:
+                            parent_span = span
+
+                        with self.tracer.start_span(
+                            name="websocket.close",
+                            service=span.service,
+                            resource=f"websocket {scope.get('path', '')}",
+                            span_type="websocket",
+                            child_of=parent_span,
+                            activate=True,
+                        ) as close_span:
+                            close_span.set_tag_str(COMPONENT, self.integration_config.integration_name)
+                            close_span.set_tag_str(SPAN_KIND, SpanKind.PRODUCER)
+
+                            client = scope.get("client")
+                            if len(client) >= 1:
+                                client_ip = client[0]
+                                close_span.set_tag_str("out.host", client_ip)
+                                try:
+                                    ipaddress.ip_address(client_ip)  # validate ip address
+                                    span.set_tag_str("network.client.ip", client_ip)
+                                except ValueError:
+                                    pass
```

Comment:
It would be nice to have a debug log when this operation fails (unless we expect this to be very noisy (ie it will always fail on some hosts).